### ISSUE TITLE
[feature request] Logic with multiple monitors

### ISSUE THREAD

**Issue by keshto** [LBCrion/sfwbar#87]
Right now I have the main bar on external monitor and a vertical bar on laptop with only taskbar (both in same config, so running one instance of sfwbar).

It would be nice to have the vertical bar that is on the laptop screen to go away if the main monitor is unplugged and thus have the main bar be on the laptop screen. plugging back in the external monitor should then go back to initial setup.

That might be too complicated, so an easier idea is to have sfwbar clone itself across all outputs (what waybar does).

**Comment by keshto** [comment]
Assuming the SetMonitor config worked, how would this accomplish the above scenario? This would put laptop bar on the laptop screen if only if the output exists right? the laptop output does not disconnect in the multi monitor setup nor when the external monitor is disconnected so that condition will always be true. The idea is to have laptop_bar disappear if the external monitor is unplugged and reappear when it is connected.

**Comment by Consolatis** [comment]
You could likely hack something together using kanshi + the `exec` directive.
Potentially including the new SIGHUP handling for sfwbar or by killing / executing a different sfwbar process.

**Comment by keshto** [comment]
Didnt know kamshi had a exec directive. TIL.
That would work thanks. 
Want to keep this open for a simple clone across all outputs (which probably should be default) feature.

**Comment by LBCrion** [comment]
Ah, I missed the fact that you're trying to hide a bar on one monitor on disconnection of another. This would be more complicated. Probably the best way would be to implement output connect/disconnect triggers. sfwbar already monitors for connect/disconnect events, so it should be easy.
We would also need a bar visibility control action, since the usual style based approach only applies to the grids within the bar (and bar show/hide logic has it's own complexities).
I think something along these lines:
```
TriggerAction "eDP-1_disconnect", BarSetVisibility "secondary_bar","hide"
TriggerAction "eDP-1_connect", BarSetVisibility "secondary_bar","show"
```
The second part of this issue (replicating the bar across multiple outputs) is much more complex and I recommend just launching multiple instances with -m parameter. Replicating the bar internally would be a bit of a nightmare, since all bars in the single process share variable and (this is the bigger problem) widget namespaces. I.e. we would either need to introduce per-bar widget namespaces (which would be a major breaking change), or allow multiple widgets share the same name (which would break per-outpu window collection actions).

**Comment by keshto** [comment]
Proposed solution could work.

Hmm yea that is challenging with the internal state. Running one bar that maps across multiple displays would be less memory than running a new instance per monitor also one does not get the (what i think) the sane default should be which is running across all outputs and then a person would configure otherwise rather than initial one bar one screen then the person would have to figure out how to launch multiple bars per display and have a config for each display or put a different -m arg per bar launch etc. Then what happens when they connect to a new display with a different name they need to start the bar with new name.

**Comment by keshto** [comment]
That would be nice

**Comment by LBCrion** [comment]
Wiring in the keyword would be the easy bit, it's the logic behind it that gets very messy. I guess we could keep per-bar widget namespaces. But that would still leave the questions of actions initiated without a parent widget, such as actions action called from SfwbarInit or TriggerAction.

**Comment by Consolatis** [comment]
Just thinking out loud, maybe something like a bar abstraction that provides functions to be called and internally just has a list of actual bars it owns. That list would usually only contain a single bar instance unless `all` is used. The abstraction would then just call the action on all instances of its internal list. I didn't really think this through but something along that line might be one option.

**Comment by LBCrion** [comment]
That would be a sensible approach. A few other things that came up in my thinking about replicating the bar:

1) state - the state of the bar can change over time through user actions or external inputs. I.e. a SetValue or SetStyle can be called on widgets, UserState's may be modified etc. How should we treat user state of a bar when an output goes away? Shell we store it in hopes that the output comes back (i.e. hide the bar, while the output is gone) or discard it and initialize the bar from scratch when the output comes back?
2) variables - scanner variables get updated whenever a widget requests the value. If we replicate the bar across multiple outputs, the widgets are replicated and start requesting variable values more often. This can cause issues with rate type variables (i.e. CPU utilization). The first copy of a widget requests CPU counters and updates with change's in these from the last poll, then the second widget requests them again and updates with differences from the first widget's call. This can cause inaccurate readings.

One approach I think may work is to allow mirroring of a bar. I.e. have a bar mapped to a single output (which updates values etc) and then mirror it to other outputs (i.e. values of widgets and state get mirrored from the primary bar). The state of the primary is persistent, if it's output goes away, it just gets hidden. 

Just to manage expectations, this feature won't be quick to implement, as it would require some fundamental changes to the internals of sfwbar, I would like to get the next version out before starting any work on it. And really would like to track down #86 before releasing the next version.

**Comment by keshto** [comment]
I like the mirroring idea since that is sorta what one would want if they say `all` they dont expect a unique bar but rather a clone of whatever they specified across all outputs.

I'm fine with this taking a back seat. #86 is for sure more important. So far, have only had sfbar freeze up for a little when I middle clicked closed a window from taskbar. Still waiting to see if I can reliably reproduce the 100% CPU spike.

**Comment by LBCrion** [comment]
I merged the initial implementation of the mirroring functionality. There are now two new actions: SetMirror and BlockMirror, which specify which outputs bar mirrors should appear on, the simplest config would be:
```
Function("SfwbarInit") {
  SetMirror "*"
}
```

This will mirror all bars to all outputs. You can mirror only a specific bar with ``SetMirror "bar","*"``` or mirror a specific bar to a specific output with ``SetMirror "bar","e-DP1"``. You can also exclude a specific output from mirroring with ``BlockMirror "e-DP2"``. You can also specify semicolon separated lists of outputs for both SetMirror and BlockMirror.

Unfortunately I had to change the behaviour of new bars to always launch on the first output rather than the current output. The bug/deficiency in gtk firmly prevents me from keeping the old behaviour.

The technical explanation: to mirror a bar, we first need to know the output of the parent bar (to avoid duplicating the bar on this output). The output of the bar is only known once the wayland client receives the ``enter`` event on the bar window surface. Unfortunately, GTK doesn't provide any signal, callback or any other type of a notification when it receives this event. To make things worse, it will always return an output associated with the window, except that if you query it before the ``enter`` event has been received, the result will be wrong (it will be based on a fallback approach which doesn't work well for wayland). As a result, in order to know the output of the. window, with any confidence, we need to set the output manually.

The only (very hacky) workaround I can think of is to create a layer shell window on start-up (not a GTK based one, but a pure wayland implementation), listen for the enter event on it's surface and once, the default output is known, destroy this window. We can then use this output as the default output for new bars. I'm not sure I want to go down this route yet, since it's rather cumbersome and has it's own drawbacks (i.e. what if the output get disconnected after we query it).

**Comment by LBCrion** [comment]
I gave the above hack a try and it works better than I expected. Since primary bars are always created on init, manually probing for default output and assigning them to it (unless explicitly specified otherwise) seems to work quite well. As a side benefit, it removes a lot of cruft around bar hide/show code and monitor jumping on output connects/disconnects.

Mirroring should now work with the main bar launching on the active output.
